### PR TITLE
Add endpoint lock Redis object sc 19081

### PR DESCRIPTION
**Pull request by LeiGlobus**
Add endpoint lock in Redis

### CODE REVIEW

**Comment by joshbryan-globus** [comment]
[sc-19081]

File: tests/unit/test_redis_task.py [hunk 1]
```diff
@@ -22,7 +22,7 @@
 
 @pytest.fixture(autouse=True)
 def lower_ttl_for_test():
-    RedisTask.DEFAULT_TTL = 10  # Play nice with local infrastructure
+    RedisTask.DEFAULT_TTL = 1000  # Play nice with local infrastructure
```
Comment (khk-globus) [inline tests/unit/test_redis_task.py:25]: What's the reason for this change?  For example, I see no change to the `RedisTask` tests.

File: src/funcx_common/response_errors/error_classes.py [hunk 2]
```diff
@@ -163,6 +163,32 @@ def __init__(self, ep_uuid: t.Union[uuid.UUID, str]) -> None:
...
+    """Endpoint with specified uuid cannot be registered until the lock expires"""
+
+    code = ResponseErrorCode.RESOURCE_LOCKED
+    http_status_code = HTTPStatusCode.RESOURCE_LOCKED
+
+    def __init__(self, ep_uuid: t.Union[uuid.UUID, str]) -> None:
+        ep_uuid = str(ep_uuid)
+        self.error_args = [ep_uuid]
+        self.reason = f"Endpoint id {ep_uuid} is temporarily locked from registration"
+        self.uuid = ep_uuid
```
Comment (khk-globus) [inline src/funcx_common/response_errors/error_classes.py:189]: I don't yet know how this will be utilized, but I can see a potential use of knowing _when_ the lock will expire.  (For example, in a response to the end-user or in the logs.)

File: src/funcx_common/response_errors/error_classes.py [hunk 3]
```diff
@@ -163,6 +163,32 @@ def __init__(self, ep_uuid: t.Union[uuid.UUID, str]) -> None:
...
+    """Endpoint with specified uuid is already registered"""
+
+    code = ResponseErrorCode.RESOURCE_CONFLICT
+    http_status_code = HTTPStatusCode.RESOURCE_CONFLICT
+
+    def __init__(self, ep_uuid: t.Union[uuid.UUID, str]) -> None:
+        ep_uuid = str(ep_uuid)
+        self.error_args = [ep_uuid]
+        self.reason = f"Endpoint id {ep_uuid} is already registered to this user"
+        self.uuid = ep_uuid
```
Comment (khk-globus) [inline src/funcx_common/response_errors/error_classes.py:176]: I'm not fully clear on our terminology, but seems to me that "InUse" is not the same thing as "registered."  As text to an end-user, I would find the phrase "registered" to be confusing in this context.  "Of course it's registered.  I registered it yesterday!"

Suggest perhaps:

> Endpoint currently in-use. (id: {ep_uuid})

File: src/funcx_common/redis_endpoint_lock.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        now_ts = datetime.now().timestamp()
+        return now_ts + self.redis_client.pttl(self.hname) / 1000
+
+    @expire_ts.setter
+    def expire_ts(self, expiration_ts: float) -> None:
+        """Expires lock after expiration time, if not already set."""
+        ttl_ms = self.redis_client.pttl(self.hname)
+        if ttl_ms < 0:
+            expiration_ts_ms = int(expiration_ts * 1000)
+            self.redis_client.pexpireat(self.hname, expiration_ts_ms)
```
Comment (khk-globus) [inline src/funcx_common/redis_endpoint_lock.py:80]: Doesn't look like these code paths have any tests?

File: src/funcx_common/redis_endpoint_lock.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            return cls(redis_client, endpoint_id)
+        raise ValueError(f"Cannot find lock {endpoint_id}")
+
+    @classmethod
+    def is_locked(cls, redis_client: "redis.Redis[t.Any]", endpoint_id: str) -> bool:
+        # We don't need the info in the stored object, just using built in
+        # Redis expiry functionality to see if there's a live lock
+        # TODO add more information to the stored object and possibly
+        # checking the stored expiry timestamp
+        return cls.exists(redis_client, endpoint_id)
```
Comment (khk-globus) [inline src/funcx_common/redis_endpoint_lock.py:108]: Seems that either `.exists()` or this method is redundant, no?  What's the compelling reason to have both?

File: tests/unit/test_redis_endpoint_lock.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    import redis
+
+    has_redis = True
+except ImportError:
+    has_redis = False
+
+if not has_redis or not LOCAL_REDIS_REACHABLE:
+    pytest.skip(
+        "these tests only run with access to local redis", allow_module_level=True
+    )
```
[the remark this answers is not in the document]
Comment (LeiGlobus) [reply tests/unit/test_redis_endpoint_lock.py]: Yes it makes sense that LOCAL_REDIS_REACHABLE checks for both, but I wasn't quite sure what different import and missing redis server paths were being tested by the original code in test_redis_task, so felt it safer to do it the same way.  But as this is done in test_redis_task anyway, I guess there's no reason to do it here too.  So simplified as you suggested.

File: src/funcx_common/redis_endpoint_lock.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    @classmethod
+    def load(
+        cls, redis_client: "redis.Redis[t.Any]", endpoint_id: str
+    ) -> "RedisEndpointLock":
+        """
+        Returns an object lock if it exists, otherwise None
+        """
+        if cls.exists(redis_client, endpoint_id):
+            return cls(redis_client, endpoint_id)
+        raise ValueError(f"Cannot find lock {endpoint_id}")
```
[the remark this answers is not in the document]
Comment (LeiGlobus) [reply src/funcx_common/redis_endpoint_lock.py]: Yes good catch.  It's a left over from when it did return None, but after the change it wasn't updated.  Updated.

File: tests/unit/test_redis_endpoint_lock.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+
+@pytest.fixture
+def redis_client():
+    return redis.Redis("localhost", port=6379, decode_responses=True)
```
[the remark this answers is not in the document]
Comment (LeiGlobus) [reply tests/unit/test_redis_endpoint_lock.py]: Changed.

File: tests/unit/test_redis_endpoint_lock.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    expiration_time = datetime.now() + timedelta(milliseconds=expiration_period_ms)
+    expiration_time_ts = expiration_time.timestamp()
+    if expiration_period_ms <= 0:
+        with pytest.raises(ValueError):
+            RedisEndpointLock(
+                redis_client,
+                endpoint_id,
+                lock_timestamp=expiration_time_ts,
+            )
+        return
```
[the remark this answers is not in the document]
Comment (LeiGlobus) [reply tests/unit/test_redis_endpoint_lock.py]: Separated.

File: tests/unit/test_redis_endpoint_lock.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    with pytest.raises(ValueError):
+        RedisEndpointLock.load(redis_client, endpoint_id)
```
[the remark this answers is not in the document]
Comment (LeiGlobus) [reply tests/unit/test_redis_endpoint_lock.py]: Refactored as this comment and below suggests.

File: tests/unit/test_redis_endpoint_lock.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    sleep(0.5)
+    if expiration_period_ms < 500:
+        assert not RedisEndpointLock.exists(redis_client, endpoint_id)
+    else:
+        assert RedisEndpointLock.exists(redis_client, endpoint_id)
+        assert RedisEndpointLock.is_locked(redis_client, endpoint_id)
+        loaded_lock = RedisEndpointLock.load(redis_client, endpoint_id)
+        assert loaded_lock.lock_expiration_ts == expiration_time_ts
+        sleep(expiration_period_ms / 1000 - 0.5)
+        assert not RedisEndpointLock.exists(redis_client, endpoint_id)
```
[the remark this answers is not in the document]
Comment (LeiGlobus) [reply tests/unit/test_redis_endpoint_lock.py]: Removed if logic.

File: tests/unit/test_redis_task.py (hunk 1 above)
Comment (LeiGlobus) [reply tests/unit/test_redis_task.py]: Originally changed to 1000 to test expiry behavior in debugger, but didn't change back.

File: src/funcx_common/response_errors/error_classes.py (hunk 2 above)
Comment (LeiGlobus) [reply src/funcx_common/response_errors/error_classes.py:189]: I was debating whether to return the lock timestamp as part of the error, but felt it was extra logic (and passing of parameters) that may not be needed for a seconds long lock.  Could add it in, but let me think and we can discuss.

File: src/funcx_common/redis_endpoint_lock.py (hunk 4 above)
Comment (LeiGlobus) [reply src/funcx_common/redis_endpoint_lock.py:80]: expire_ts is used to actually expire, so all tests use the setter.  The getter for expire_ts is the same as lock_expiration_ts, so thought was too simple logic to test.  I suppose I can add a line to ensure it's equal if you really think it's useful.

File: src/funcx_common/redis_endpoint_lock.py (hunk 5 above)
Comment (LeiGlobus) [reply src/funcx_common/redis_endpoint_lock.py]: I used is_locked as a wrapper because the wording is more clear.  But didn't remove the exists() because it's a more generic method that all redis classes probably should have.  It just happens that exists() and is_locked() is the same for this class, but maybe in the future we have existing lock objects that are expired or in some other state?

But as they are the same as of now, removed is_locked()

File: tests/unit/test_redis_endpoint_lock.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (LeiGlobus) [reply tests/unit/test_redis_endpoint_lock.py]: Addendum - the try/import/except block was necessary, not just 'import redis' as the 3.7 CI test failed to import redis and threw an exception.  Then the 'not has_redis or not LOCAL...' was just so it doesn't do a pytest.skip() twice.  See modified code, let me know which way is better.

File: src/funcx_common/response_errors/error_classes.py (hunk 2 above)
Comment (LeiGlobus) [reply src/funcx_common/response_errors/error_classes.py]: lock timestamp added.

File: tests/unit/test_redis_endpoint_lock.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (khk-globus) [reply tests/unit/test_redis_endpoint_lock.py]: Either really; not meant to get stuck on this.  Just a suggestion.  There "just" tests.

**Review by khk-globus — APPROVED** [review]
Nice!
